### ISSUE TITLE
[java] Rules for detecting overload methods in Java

### ISSUE THREAD

**Issue by caesarkim** [pmd/pmd#2768]
**Description:**

I am looking into a rule that can detect overload methods in Java. I tried it with PMD Rule Designer which shows the AST and XPath.  I don't think that I can find a solution for that, so I am wondering if there is a way to do it, even writing a custom java rule.

Here is my simple usecase.  Even though it is syntactically incorrect, please understand.

```java
public class ExampleFuture {

  public static SampleFuture delay(final Runnable action, final long delay) {
    return new SampleFuture(action, delay, null);
  }

  public static <T> SampleFuture<T> delay(final Supplier<T> action, final long delay) {
    return new SampleFuture(action, delay, TimeUnit.MILLISECONDS, null);
  }
}


public class ExampleExecution {

  public static void execute() {
    ExampleFuture .delay(() -> {
      System.out.println("Hello world");
    }, 100L);
  }
}
```

This is a simplified version of class that I have. The class 'ExampleExecution' is calling 'ExampleFuture .delay' which has Runnable and long types for arguments.  I want to detect 'public static SampleFuture delay(final Runnable action, final long delay)' usage in the ExampleExecution.   This is just one of my usecases. It could be overload methods with various object types. 

When I see this code in the PMD Rule Designer, I don't see types of arguments which are passed into 'ExampleFuture .delay'. I guess that it is because PMD analyzes at the code level instead of runtime level. 

I am wondering if there is a way to do it even in the custom java rule, if it is not supported in the PMD Rule Designer. 

Thanks.

**Comment by oowekyala** [comment]
For now this is not supported on master, sorry. We don't preserve the result of overload selection, apart from the final type of the method call. So a rule cannot know what overload between these two is being called... Fwiw, there will be an API for that in PMD 7 (#2689)
